helix-view/src/modeline.rs

```diff
@@ -0,0 +1,258 @@
+
+#[derive(Default, Debug, Eq, PartialEq)]
+pub struct Modeline {
+    language: Option<String>,
+    indent_style: Option<IndentStyle>,
+    line_ending: Option<LineEnding>,
+}
+
+impl Modeline {
+    pub fn parse(text: &Rope) -> Self {
+        let mut modeline = Self::default();
+
+        for line in text.lines().take(LINES_TO_CHECK).chain(
+            text.lines_at(text.len_lines())
+                .reversed()
+                .take(LINES_TO_CHECK),
+        ) {
+            // can't guarantee no extra copies, since we need to regex and
+            // regexes can't operate over chunks yet, but we can at least
+            // limit how much we potentially need to copy because modelines
+            // are typically quite short.
+            let line = Cow::<str>::from(line.slice(..(LENGTH_TO_CHECK.min(line.len_chars()))));
+            if let Some(line_modeline) = Self::parse_line(&line) {
+                modeline.merge(line_modeline);
+            }
+        }
+
+        modeline
+    }
+
+    pub fn language(&self) -> Option<&str> {
+        self.language.as_deref()
+    }
+
+    pub fn indent_style(&self) -> Option<IndentStyle> {
+        self.indent_style
+    }
+
+    pub fn line_ending(&self) -> Option<LineEnding> {
+        self.line_ending
+    }
+
+    fn parse_line(line: &str) -> Option<Self> {
+        if let Some(captures) = MODELINE_REGEX.captures(line) {
+            let mut language = None;
+            let mut indent_style = None;
+            let mut line_ending = None;
+
+            let options = &captures[1];
+            for option in options.split([' ', '\t', ':']) {
+                let parts: Vec<_> = option.split('=').collect();
+                match parts[0] {
+                    "ft" | "filetype" => {
+                        if let Some(val) = parts.get(1) {
+                            language = Some(vim_ft_to_helix_language(val));
+                        }
+                    }
+                    "sw" | "shiftwidth" => {
+                        if let Some(val) = parts.get(1).and_then(|val| val.parse().ok()) {
+                            if indent_style != Some(IndentStyle::Tabs) {
+                                indent_style = Some(IndentStyle::Spaces(val));
+                            }
+                        }
+                    }
+                    "ff" | "fileformat" => {
+                        if let Some(val) = parts.get(1) {
+                            line_ending = vim_ff_to_helix_line_ending(val);
+                        }
+                    }
+                    "noet" | "noexpandtab" => {
+                        indent_style = Some(IndentStyle::Tabs);
+                    }
+                    _ => {}
+                }
+            }
+
+            return Some(Self {
+                language,
+                indent_style,
+                line_ending,
+            });
+        }
+        None
+    }
+
+    fn merge(&mut self, other: Self) {
```

Comment:
I think this is unnecessary. Instead of Creating a new `ModeLine` for each line that is parsed simply make the parsing accept a mutable reference to self and write the results directly into self